presto-orc/src/main/java/com/facebook/presto/orc/writer/CompressionBufferPool.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.writer;
+
+import org.openjdk.jol.info.ClassLayout;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+
+public interface CompressionBufferPool
+{
+    byte[] checkOut(int length);
+
+    void checkIn(byte[] buffer);
+
+    long getRetainedBytes();
+
+    class LastUsedCompressionBufferPool
+            implements CompressionBufferPool
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(LastUsedCompressionBufferPool.class).instanceSize();
+        private byte[] lastUsed;
+
+        @Override
+        public byte[] checkOut(int length)
+        {
+            if (lastUsed == null || lastUsed.length < length) {
+                return new byte[length];
+            }
+            byte[] returnValue = lastUsed;
+            lastUsed = null;
+            return returnValue;
+        }
+
+        @Override
+        public void checkIn(byte[] buffer)
+        {
+            lastUsed = buffer;
```

Comment:
1. There is a probability that multiple buffers will be returned back to pool (even if we don't do it now). Do you want to add any logic on choosing wether to keep current lastUsed or replace it with the returned buffer? This will add a bit of stability to this pool.
2. requireNonNull?

Reply:
Btw, there is such thing as com.facebook.presto.orc.stream.SharedBuffer. May be we can wrap the pool around it instead of raw buffers.

Reply:
I looked at the SharedBuffer, SharedBuffer is tied with memoryContext, so it would require some refactoring before we could use that. But what SharedBuffer did for the reader, this change is doing for the writer. Thanks for pointing out that.

About multiple buffers and thread safety, I originally wrote code using AtomicReference and thread safety and handle multiple requests gracefully. But most of the OrcWriter will be busted if we go towards multiple threads, so making just this part complicated did not make sense, so I removed all of those in the end.

Reply:
Right, writer is not thread safe, I don't have concerns here. 

But you know what, what if ColumnWriterOptionsBuilder is reused? In this case the pool created by builder constructor will be shared among different writers. Let's create the pool in the Builder.build method to make sure every build() call gets its own pool.

Reply:
OrcWriter creates the ColumnWriterOptionsBuilder and it is not externally exposed, so that should not be a problem in my opinion.  I first put in CompressionBufferPool in OrcWriterOptions (which can be shared) then removed it for this reason.

there could be a problem with unit tests reusing the ColumnWriterOptionsBuilder across the tests, but that will be a test issue.